# [47장] 에러 처리

### 47.1 에러 처리의 필요성
- 발생한 에러에 대해 대처하지 않고 방치하면 프로그램은 강제 종료
- exception: 직접적으로 에러를 발생하지는 않지만 예외적인 상황에 적절하게 대응하지 않으면 에러로 이어질 가능성이 큼

### 47.2 try...catch...finally 문
#### 에러 처리를 구현하는 방법
1. 예외적인 상황이 발생하면 if 문이나 단축 평가 또는 옵셔널 체이닝 연산자를 통해 확인해서 처리하는 방법
2. 에러 처리 코드를 미리 등록해 두고 에러가 발생하면 에러 처리 코드로 점프하도록 하는 방법 → `try...catch...finally` 문
  ```
  try {
    // 실행할 코드
  } catch (err) {
    // try 코드 블록에서 에러 발생 시 실행
  } finally {
    // 에러 발생과 상관없이 반드시 한 번 실행 → 생략 가능
  }
  ```

### 47.3 Error 객체
- Error 생성자 함수는 에러 객체 생성
- 에러를 상세히 설명하는 에러 메시지를 인수로 전달할 수 있음
```
const error = new Error('invalid!');
```
생성자 함수 | 인스턴스
--|--
Error | 일반적 에러 객체
SyntaxError | js 문법에 맞지 않는 문을 해석할 때 발생하는 에러 객체
ReferenceError | 참조할 수 없는 식별자를 참조했을 때 발생하는 에러 객체
TypeError | 피연산자 또는 인수의 데이터 타입이 유효하지 않을 때 발생하는 에러 객체
RangeError | 숫자값의 허용 범위를 벗어났을 때 발생하는 에러 객체
URIError | encodeURI 또는 decodeURI 함수에 부적절한 인수를 전달했을 때 발생하는 에러 객체
EvalError | eval 함수에서 발생하는 에러 객체

### 47.4 throw 문
에러를 발생시키려면 try 코드 블록에서 throw 문으로 에러 객체를 던져야 함 → catch 코드 블록 실행

`throw 표현식;`

### 47.5 에러의 전파
- throw된 에러를 캐치하지 않으면 호출자 방향으로 전파됨
- 어디에서도 catch하지 않으면 프로그램 강제 종료
- 주의) setTimeout이나 프로미스 후속 처리 메서드의 콜백 함수는 호출자가 없음 → 마이크로태스크 큐에 일시 저장되었다가 콜 스택으로 이동하면서 콜 스택의 가장 하부에 존재하기 때문